src/compiler/factory.ts

```diff
@@ -2629,25 +2629,130 @@ namespace ts {
     export function createUnparsedSourceFile(inputFile: InputFiles, type: "js" | "dts"): UnparsedSource;
     export function createUnparsedSourceFile(text: string, mapPath: string | undefined, map: string | undefined): UnparsedSource;
     export function createUnparsedSourceFile(textOrInputFiles: string | InputFiles, mapPathOrType?: string, map?: string): UnparsedSource {
         const node = <UnparsedSource>createNode(SyntaxKind.UnparsedSource);
         if (!isString(textOrInputFiles)) {
             Debug.assert(mapPathOrType === "js" || mapPathOrType === "dts");
-            node.fileName = mapPathOrType === "js" ? textOrInputFiles.javascriptPath : textOrInputFiles.declarationPath;
+            node.fileName = (mapPathOrType === "js" ? textOrInputFiles.javascriptPath : textOrInputFiles.declarationPath) || "";
             node.sourceMapPath = mapPathOrType === "js" ? textOrInputFiles.javascriptMapPath : textOrInputFiles.declarationMapPath;
             Object.defineProperties(node, {
                 text: { get() { return mapPathOrType === "js" ? textOrInputFiles.javascriptText : textOrInputFiles.declarationText; } },
                 sourceMapText: { get() { return mapPathOrType === "js" ? textOrInputFiles.javascriptMapText : textOrInputFiles.declarationMapText; } },
             });
         }
         else {
+            node.fileName = "";
             node.text = textOrInputFiles;
             node.sourceMapPath = mapPathOrType;
             node.sourceMapText = map;
         }
+        const text = node.text;
+        node.languageVersion = ScriptTarget.Latest;
+
+        // Shebang
+        let pos = isShebangTrivia(text, 0) ? skipTrivia(text, 0, /*stopAfterLineBreak*/ true) : 0;
+
+        // Prologue
+        const scanner = createScanner(ScriptTarget.Latest, /*skipTrivia*/ true, /*languageVariant*/ undefined, text, /*onError*/ undefined, pos);
+        let prologues: UnparsedPrologue[] | undefined;
+        while (scanner.scan()  === SyntaxKind.StringLiteral) {
+            const start = pos;
+            const prologueText = scanner.getTokenValue();
+            scanner.tryScan(() => scanner.scan() === SyntaxKind.SemicolonToken);
+            pos = skipTrivia(text, scanner.getTextPos(), /*stopAfterLineBreak*/ true);
+
+            const prologue = <UnparsedPrologue>createNode(SyntaxKind.UnparsedPrologue, start, pos);
+            prologue.parent = node;
+            prologue.text = prologueText;
+            (prologues || (prologues = [])).push(prologue);
+        }
+
+        // Helpers
+        let helpers: UnscopedEmitHelpers[] | undefined;
+        while (true) {
+            const helperInfo = tryGetUnscopedEmitHelper(text, pos);
+            if (!helperInfo) break;
+            pos = helperInfo.newPos;
+            (helpers || (helpers = [])).push(helperInfo.helper);
+        }
+
+        // triple slash directives
+        const newPos = processCommentPragmas(node as {} as PragmaContext, text);

```

Comment:
Shouldn't `UnparsedSource` just inherit from `PragmaContext`, so this cast is unneeded?

Reply:
`PragmaContext` is internal but `UnparsedSource` is not. Same thing is done for `SourceFile` as well.

Reply:
Also note that #29630 is another version of this PR where we store all this information in the bundleInfo which we can also use to create internal and comment sections and handle that in the generating output depending on removeComments and --stripInternal in the project output.